src/main/java/org/mars/rover/kata/MarsNavigator.java

```diff
@@ -0,0 +1,83 @@
+package org.mars.rover.kata;
+
+import java.util.Scanner;
+import java.util.ArrayList;
+
+public class MarsNavigator {
+    Scanner scanner;
+    ArrayList<String> providedInput;
+
+    ArrayList<ArrayList<Coordinate>> grid;
+
+    ArrayList<MarsRover> marsRovers;
+
+    public MarsNavigator(Scanner scanner) {
+        this.scanner = scanner;
+        this.providedInput = new ArrayList<>();
+        this.grid = new ArrayList<>();
+        this.marsRovers = new ArrayList<>();
+    }
+
+    public void loadInput() {
+        // get the input from the scanner
+        while (this.scanner.hasNext()) {
+            this.providedInput.add(this.scanner.nextLine());
+        }
+    }
+
+    public void processInput() {
+        this.constructCoordinates();
+        this.constructRovers();
+    }
+
+    protected void constructCoordinates() {
+        // get first input so we can construct the coordinates
+        String[] parts = this.providedInput.get(0).split(" ");
+
+        int x = Integer.parseInt(parts[0]);
+        int y = Integer.parseInt(parts[1]);
+
+        for (int i = 0; i < x; i++) {
+            ArrayList<Coordinate> row = new ArrayList<>();
+
+            for (int z = 0; z < y; z++) {
+                row.add(new Coordinate(i, z));
+            }
+
+            grid.add(row);
+        }
+    }
+
+    public void constructRovers() {
+        for (int i = 1; i < providedInput.size(); i += 2) {
+            try {
+                String[] partsRover = providedInput.get(i).split(" ");
+                int positionX = Integer.parseInt(partsRover[0]);
+                int positionY = Integer.parseInt(partsRover[1]);
+                // String roverDirection = partsRover[2];
+
+                MarsRover marsRover = new MarsRover();
+
+                marsRover.setDirection(Direction.valueOf(partsRover[2]));
+                marsRover.setCoordinate(grid.get(positionX).get(positionY));
+
+                marsRovers.add(new MarsRover());
+                // providedInput.get(i + 1); todo implement
+            } catch (IndexOutOfBoundsException ignored) {
```

Comment:
No właśnie. Moje doświadczenie podpowiada mi, że takie rzeczy zostają ignorowane już na zawsze w projektach.

Dlatego jak piszę nowe funkcje w programie, to zaczynam zawsze pisać testy od negatywnych przypadków. Typu dane wejściowe są niepoprawne, nie da się ich sparsować itd. W ten sposób od razu na początku oprogramowuję sytuacje wyjątkowe, nazywam je biznesowo (np. wyjątek o nazwie `CannotMoveOutsideOfGrid` zamiast generyczne `IndexOutOfBoundException`).

Reply:
Dodane.